### ISSUE TITLE
avahi stopping after sometime

### ISSUE THREAD

**Issue by vajonam** [DrPsychick/docker-cups-airprint#39]
I have an issue where the avahi-daemon stops after sometime. and restarting the container doesn't fix it. To fix it I have to shell into the box and restart avahi-daemon. Have you seen this behavior? or maybe add a automatic re starter for avahi?

**Comment by vajonam** [comment]
```
Setting up watches.
Watches established.
Found user 'avahi' (UID 105) and group 'avahi' (GID 111).
Successfully dropped root privileges.
avahi-daemon 0.7 starting up.
Successfully called chroot().
Successfully dropped remaining capabilities.
Loading service file /services/AirPrint-HP_Color_Laserjet_2600n.service.
Joining mDNS multicast group on interface eth0.IPv4 with address 192.168.100.14.
New relevant interface eth0.IPv4 for mDNS.
Joining mDNS multicast group on interface lo.IPv4 with address 127.0.0.1.
New relevant interface lo.IPv4 for mDNS.
Network interface enumeration completed.
Registering new address record for 192.168.100.14 on eth0.IPv4.
Registering new address record for 127.0.0.1 on lo.IPv4.
Server startup complete. Host name is cups.local. Local service cookie is 201674687.
E [30/Jun/2022:18:15:51 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:51 +0000] Bad Listen address \\*:631 at line 128.
E [30/Jun/2022:18:15:51 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:51 +0000] Bad Listen address \\*:631 at line 129.
E [30/Jun/2022:18:15:51 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:51 +0000] Bad Listen address \\*:631 at line 130.
E [30/Jun/2022:18:15:51 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:51 +0000] Bad Listen address \\*:631 at line 131.
E [30/Jun/2022:18:15:51 +0000] Unable to open listen socket for address [v1.::]:631 - Address family not supported by protocol.
E [30/Jun/2022:18:15:51 +0000] Unable to communicate with avahi-daemon: Daemon not running
Service "AirPrint HP_Color_Laserjet_2600n @ cups" (/services/AirPrint-HP_Color_Laserjet_2600n.service) successfully established.
localhost - - [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 401 0 - -
localhost - root [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 201 4712 - -
.--> CUPS ready
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 128.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 129.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 130.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 131.
E [30/Jun/2022:18:15:52 +0000] Unable to open listen socket for address [v1.::]:631 - Address family not supported by protocol.
E [30/Jun/2022:18:15:52 +0000] Unable to communicate with avahi-daemon: Daemon not running
localhost - - [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 401 0 - -
localhost - root [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 201 4712 - -
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 128.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 129.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 130.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 131.
E [30/Jun/2022:18:15:52 +0000] Unable to open listen socket for address [v1.::]:631 - Address family not supported by protocol.
E [30/Jun/2022:18:15:52 +0000] Unable to communicate with avahi-daemon: Daemon not running
localhost - - [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 401 0 - -
localhost - root [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 201 4712 - -
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 128.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 129.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 130.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 131.
E [30/Jun/2022:18:15:52 +0000] Unable to open listen socket for address [v1.::]:631 - Address family not supported by protocol.
E [30/Jun/2022:18:15:52 +0000] Unable to communicate with avahi-daemon: Daemon not running
localhost - - [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 401 0 - -
localhost - root [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 201 4712 - -
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 128.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 129.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 130.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 131.
E [30/Jun/2022:18:15:52 +0000] Unable to open listen socket for address [v1.::]:631 - Address family not supported by protocol.
E [30/Jun/2022:18:15:52 +0000] Unable to communicate with avahi-daemon: Daemon not running
localhost - - [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 401 0 - -
localhost - root [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 201 4712 - -
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 128.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 129.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 130.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 131.
E [30/Jun/2022:18:15:52 +0000] Unable to open listen socket for address [v1.::]:631 - Address family not supported by protocol.
E [30/Jun/2022:18:15:52 +0000] Unable to communicate with avahi-daemon: Daemon not running
localhost - - [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 401 0 - -
localhost - root [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 201 4712 - -
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 128.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 129.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 130.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 131.
E [30/Jun/2022:18:15:52 +0000] Unable to open listen socket for address [v1.::]:631 - Address family not supported by protocol.
E [30/Jun/2022:18:15:52 +0000] Unable to communicate with avahi-daemon: Daemon not running
localhost - - [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 401 0 - -
localhost - root [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 201 4712 - -
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 128.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 129.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 130.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 131.
E [30/Jun/2022:18:15:52 +0000] Unable to open listen socket for address [v1.::]:631 - Address family not supported by protocol.
E [30/Jun/2022:18:15:52 +0000] Unable to communicate with avahi-daemon: Daemon not running
localhost - - [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 401 0 - -
localhost - root [30/Jun/2022:18:15:52 +0000] "PUT /admin/conf/cupsd.conf HTTP/1.1" 201 4712 - -
--> adding printers
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 128.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 129.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 130.
E [30/Jun/2022:18:15:52 +0000] Hostname lookup for "\\*" failed.
E [30/Jun/2022:18:15:52 +0000] Bad Listen address \\*:631 at line 131.
E [30/Jun/2022:18:15:52 +0000] Unable to open listen socket for address [v1.::]:631 - Address family not supported by protocol.
E [30/Jun/2022:18:15:52 +0000] Unable to communicate with avahi-daemon: Daemon not running
CUPS_LPADMIN_PRINTER1 = lpadmin -p "HP_Color_Laserjet_2600n" -D "HP Color Lasejet 2600n" -L "Upstaris Office" -E -v socket://192.168.100.13 -o ColorMode=Color -o InputSlot=Auto -o PageSize=Letter -o pdftops-renderer-default=pdftops -m foo2zjs:0/ppd/foo2zjs/HP-Color_LaserJet_2600n.ppd
lpadmin: Printer drivers are deprecated and will stop working in a future version of CUPS.
localhost - - [30/Jun/2022:18:15:52 +0000] "POST / HTTP/1.1" 200 43617 CUPS-Get-PPD -
localhost - - [30/Jun/2022:18:15:52 +0000] "POST /admin/ HTTP/1.1" 401 43776 CUPS-Add-Modify-Printer successful-ok
localhost - root [30/Jun/2022:18:15:52 +0000] "POST /admin/ HTTP/1.1" 200 43776 CUPS-Add-Modify-Printer successful-ok
--> CUPS configured
===========================================================
The dockerized CUPS instance is now ready for use! The web
interface is available here:
URL:       http://192.168.100.14:631/ or http://cups.manojav.com:631/
Username:  admin
Password:  admin

Google Cloud Print
local: false
cloud: false
===========================================================
Files changed, reloading.
Service group file /services/AirPrint-HP_Color_Laserjet_2600n.service vanished, removing services.
No service file found in /etc/avahi/services.
Files changed, reloading.
Loading service file /services/AirPrint-HP_Color_Laserjet_2600n.service.
Service "AirPrint HP_Color_Laserjet_2600n @ cups" (/services/AirPrint-HP_Color_Laserjet_2600n.service) successfully established.
```

**Comment by DrPsychick** [comment]
Actually I have the same error, but it works and avahi is running:
```
I [16/Jun/2022:16:52:06 +0000] Listening to 0.0.0.0:631 (IPv4)
I [16/Jun/2022:16:52:06 +0000] Listening to [v1.::]:631 (IPv6)
I [16/Jun/2022:16:52:06 +0000] Listening to /run/cups/cups.sock (Domain)
I [16/Jun/2022:16:52:06 +0000] Remote access is enabled.
I [16/Jun/2022:16:52:06 +0000] Loaded configuration file "/etc/cups/cupsd.conf"
I [16/Jun/2022:16:52:06 +0000] Configured for up to 100 clients.
I [16/Jun/2022:16:52:06 +0000] Allowing up to 100 client connections per host.
I [16/Jun/2022:16:52:06 +0000] Using policy "default" as the default.
I [16/Jun/2022:16:52:06 +0000] Partial reload complete.
I [16/Jun/2022:16:52:06 +0000] Listening to 0.0.0.0:631 on fd 4...
I [16/Jun/2022:16:52:06 +0000] Listening to [v1.::]:631 on fd 5...
I [16/Jun/2022:16:52:06 +0000] Listening to /run/cups/cups.sock on fd 6...
I [16/Jun/2022:16:52:06 +0000] Resuming new connection processing...
E [16/Jun/2022:16:52:06 +0000] Unable to communicate with avahi-daemon: Daemon not running
[...]
Files changed, reloading.
Loading service file /services/AirPrint-Epson-RX520.service.
Service "AirPrint Epson-RX520 @ cups" (/services/AirPrint-Epson-RX520.service) successfully established.
```

The last lines seems to come from avahi, detecting the service and loading it.

**Comment by smilence86** [comment]
I faced this issue yet, My cups running in openwrt via docker at a router, So I add a cron task in openwrt: 
```
0 6 * * * docker exec -it cups service avahi-daemon restart
```
The avahi service will restart every morning 6 o'clock.

**Comment by vajonam** [comment]
This is my solution. similar I think. restarts every hour. 

`0 */1 * * * root docker exec cupsd /opt/airprint/airprint-generate.py -d /etc/avahi/services`

**Comment by DrPsychick** [comment]
@vajonam if this fixes your issue, then there could be a problem with `inotifywait`, because the cronjob you posted does not restart avahi, it just regenerates the services from the `printers.conf`. Can you check if `inotifywait -m -e close_write,moved_to,create /etc/cups` works in your container or throws an error? If it does not exit, try `touch /etc/cups/printers.conf` in another shell and see if you get output. If so, it should work fine.

Unfortunately, the container is a "normal" container without systemd services, so running a cronjob is not possible as-is, so I created a PR to run a loop that ensures avahi is running. Let me know if that works @smilence86 : https://github.com/DrPsychick/docker-cups-airprint/pull/40

**Comment by vajonam** [comment]
That `inotifywait` command works as intended waits and get a response when I touch the file. 

I like the PR of running avahi in loop, I was thinking of doing some sort of watchdog or using supervisord like some other images. This will work for now. 

I am getting a airprint enabled printer soon so this will no-longer be my default printer so less of an issue for me going forward thanks for looking into it tho.

**Comment by smilence86** [comment]
> @vajonam if this fixes your issue, then there could be a problem with `inotifywait`, because the cronjob you posted does not restart avahi, it just regenerates the services from the `printers.conf`. Can you check if `inotifywait -m -e close_write,moved_to,create /etc/cups` works in your container or throws an error? If it does not exit, try `touch /etc/cups/printers.conf` in another shell and see if you get output. If so, it should work fine.
> 
> Unfortunately, the container is a "normal" container without systemd services, so running a cronjob is not possible as-is, so I created a PR to run a loop that ensures avahi is running. Let me know if that works @smilence86 : #40

Yeah, I think it's worked for me, My host machine is a arm 64 bit router (nanopi r2c) running openwrt, It could run cronjob with root user as default, and logs shows that's did every day.

**Comment by DrPsychick** [comment]
I merged the PR, hope that helps some ppl. with avahi dying or not running.

**Comment by DrPsychick** [comment]
It is, however the `latest` build does not go through as `google-cloud-print-connector` does not exist for the latest Ubuntu image (yet). Try with `focal`

**Comment by DrPsychick** [comment]
Ok, cloud-print is actually EOL, so I've created a PR to remove it....

**Comment by DrPsychick** [comment]
fixed with #42
